File: joyaapi/api/src/infrastructure/controllers/sensor/sensor.ts
import { ApiProperty } from '@nestjs/swagger';
import {
  IsBoolean,
  IsNotEmpty,
  IsNumber,
  IsOptional,
  IsString,
} from 'class-validator';
import {
  ICreateSensorDTO,
  IUpdateSensorAlertsDTO,
  IUpdateSensorDTO,
} from '../../../domain/ports/sensor.port';
import { Types } from 'mongoose';

abstract class CreateSensorDTO implements ICreateSensorDTO {
  @ApiProperty()
  @IsString()
  @IsNotEmpty()
  serial_number: string;
}

abstract class UpdateSensorDTO implements IUpdateSensorDTO {
  @ApiProperty()
  @IsString()
  @IsNotEmpty()
  plant_id: Types.ObjectId;

  @ApiProperty()
  @IsString()
  @IsNotEmpty()
  name: string;

  @ApiProperty()
  @IsString()
  @IsOptional()
  location: string;
}

abstract class UpdateSensorAlertsDTO implements IUpdateSensorAlertsDTO {
  @ApiProperty()
  @IsNumber()
  @IsOptional()
  humidity_alert: number;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  humidity_alert_enabled: boolean;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  humidity_alert_sent: boolean;

  @ApiProperty()
  @IsNumber()
  @IsOptional()
  temperature_alert: number;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  temperature_alert_enabled: boolean;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  temperature_alert_sent: boolean;

  @ApiProperty()
  @IsNumber()
  @IsOptional()
  luminosity_alert: number;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  luminosity_alert_enabled: boolean;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  luminosity_alert_sent: boolean;

  @ApiProperty()
  @IsNumber()
  @IsOptional()
  soil_fertillity_alert: number;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  soil_fertillity_alert_enabled: boolean;
  @ApiProperty()
  @IsBoolean()
  @IsOptional()
  soil_fertillity_alert_sent: boolean;
}

export { CreateSensorDTO, UpdateSensorDTO, UpdateSensorAlertsDTO };

File: joya/lib/data/dto/sensor.dart
class CreateSensorDTO {
  final String? location;
  final String name;
  final String serial_number;
  final String plant_id;

  CreateSensorDTO(
      {this.location,
      required this.name,
      required this.serial_number,
      required this.plant_id});
}


class UpdateSensorAlertsDTO {
  final String serial_number;

 final int humidity_alert;
final  bool humidity_alert_enabled;


 final int temperature_alert;
 final bool temperature_alert_enabled;

 final int soil_fertillity_alert;
 final bool soil_fertillity_alert_enabled;

 final int luminosity_alert;
 final bool luminosity_alert_enabled;


 UpdateSensorAlertsDTO(
      {
        required this.serial_number,
        required this.humidity_alert,
        required this.humidity_alert_enabled,
        required this.luminosity_alert,
        required this.luminosity_alert_enabled,
        required this.soil_fertillity_alert,
        required this.soil_fertillity_alert_enabled,
        required this.temperature_alert,
        required this.temperature_alert_enabled,
       });
}
